model/modelindexer/indexer.go

```diff
@@ -265,6 +295,19 @@ func (i *Indexer) processEvent(ctx context.Context, event *model.APMEvent) error
 	atomic.AddInt64(&i.eventsAdded, 1)
 	atomic.AddInt64(&i.eventsActive, 1)
 
+	if serviceName := event.Service.Name + "." + event.Service.Version; serviceName != "." {
+		s, ok := i.services[serviceName]
+		if !ok {
+			labels := []apm.MetricLabel{
+				{Name: "service.name", Value: event.Service.Name},
+				{Name: "service.version", Value: event.Service.Version},
+			}
+			s = service{val: 0, labels: labels}
+		}
+		atomic.AddInt64(&s.val, 1)
+		i.services[serviceName] = s
```

Comment:
This appears to be growing indefinitely. What's stopping APM Server from exhausting all memory, due to many unique service names?

Reply:
That's a question I don't really have an answer to. In the same way a user can cause a cardinality explosion, if they're ingesting events from an arbitrarily large number of service name/version pairs, the same will happen here. These values could be cleaned out after some amount of inactivity, or use an LRU cache, but I'm not sure which would be the right option here.

Reply:
Could we take a similar approach to the metrics aggregations, such as txmetrics? i.e.

- have an upper bound on the service cardinality (say, 10000 unique service name/version pairs)
- report deltas only (the number of events seen for the service in the aggregation period), so we can clear them after reporting